Mlem/Models/Content/Community/Community List/CommunityListModel.swift

```diff
@@ -205,28 +117,125 @@ class CommunityListModel: ObservableObject {
+                communities: .init()
+            )
+        }
+        newAllSections.append(topSection)
         
-        // update our published value for the view to render
-        mainQueue.schedule { [weak self] in
-            self?.communities = sorted
+        let favoritesSection = withDependencies(from: self) {
+            CommunityListSection(
+                viewId: "favorites",
+                sidebarEntry: .init(sidebarLabel: nil, sidebarIcon: "star.fill"),
+                inlineHeaderLabel: "Favorites",
+                accessibilityLabel: "Favorited Communities",
+                communities: favorited
+            )
         }
+        newAllSections.append(favoritesSection)
+        if !favorited.isEmpty {
+            newVisibleSections.append(favoritesSection)
+        }
+        
+        let alphabeticSections = alphabeticSections()
+        newAllSections.append(contentsOf: alphabeticSections)
+        newVisibleSections.append(contentsOf: alphabeticSections.filter { section in
+            !section.communities.isEmpty
+        })
+        
+        return (all: newAllSections, visible: newVisibleSections)
+    }
+    
+    private func alphabeticSections() -> [CommunityListSection] {
+        let alphabetSet = Set([String].alphabet)
+        let alphabet: [String] = .alphabet
+        
+        var communities: [String: [APICommunity]] = .init()
+        communities["other"] = .init()
+        communities[alphabet[0]] = .init()
+        
+        let sortedCommunities = subscribed.sorted()
+        var currentSectionIndex = 0
+        var currentSectionTitle: String = alphabet[0]
+        
+        // iterate through sorted communities, building up each letter's section
+        // it's not guaranteed that non-alphabetics be sorted to one side or the other of the alphabetics, so each element does have to be individually checked, hence the quick-lookup alphabetSet. They will be grouped, however, so branch prediction ought to make the performance impact of the conditional negligible
+        for community in sortedCommunities {
+            assert(community.name.first != nil, "\(community.name) has no first character!")
+            if !alphabetSet.contains(String(community.name.first!.uppercased())) {
+                assert(communities.keys.contains("other"), "No 'other' key in communities!")
+                communities["other"]?.append(community)
+            } else {
+                while currentSectionIndex < 25, !community.name.uppercased().starts(with: currentSectionTitle) {
+                    currentSectionIndex += 1
+                    currentSectionTitle = alphabet[currentSectionIndex]
+                    communities[currentSectionTitle] = .init()
+                }
+                assert(communities.keys.contains(currentSectionTitle), "No '\(currentSectionTitle)' key in communities!")
+                communities[currentSectionTitle]?.append(community)
+            }
+        }
```

Comment:
> // it's not guaranteed that non-alphabetics be sorted to one side or the other of the alphabetics, so each element does have to be individually checked, hence the quick-lookup alphabetSet

An alternative would be to use custom sorting logic, as shown below. This is slightly faster (~6ms vs ~21ms with a large number of communities) and results in less code. Lemmy only allows lowercase community names, which could have resulted in even faster solutions but unfortunately KBin allows uppercase names :(

```suggestion
        var communities: [String: [APICommunity]] = .init()
        
        var sortedCommunities = subscribed.sorted(by: { item1, item2 in
            // The "~" character has a greater value than all letters in the ASCII table, so prepending it ensures that the community is sorted to the end of the array
            let val1 = (item1.name.first?.isLetter ?? false) ? item1.name.lowercased() : "~" + item1.name.lowercased()
            let val2 = (item2.name.first?.isLetter ?? false) ? item2.name.lowercased() : "~" + item2.name.lowercased()
            return val1 < val2
        })
     
        for section in [String].alphabet {
            communities[section] = Array(sortedCommunities.prefix(
                while: { $0.name.first?.uppercased() == section }
            ))
            sortedCommunities.removeFirst(communities[section]?.count ?? 0)
        }
    
        communities["other"] = sortedCommunities
```

Reply:
Implemented 👍